### PR TITLE
test(NODE-6438): check that BSON undefined is returned from deserialize

### PR DESCRIPTION
**Pull request by nbbeeken**
#### What is changing?

- Prevent deserialize from changing the value returned for BSON undefined
- Reorganize tests around ignoreUndefined

##### Is there new documentation needed for these changes?

Yes

#### What is the motivation for this change?

I noticed no test fails if I change `BSON_DATA_UNDEFINED` handling, seems worthy of coverage given how the serializer special cases that same value.

### Double check the following

- [ ] Ran `npm run check:lint` script
- [ ] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [ ] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [ ] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

File: test/node/bson_undefined.test.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    it('returns a javascript undefined value', () => {
+      const doc = BSON.deserialize(bsonDocWithUndefined);
+      expect(doc).to.have.own.property('a').that.is.undefined;
+    });
+  });
+
+  describe('when serialize is given a javascript object that contains undefined', () => {
+    describe('when ignoreUndefined is set to false', function () {
+      it('serializes to document with a set to BSON null (type=10)', () => {
+        const jsObject = BSON.deserialize(bsonDocWithUndefined);
```
Comment (baileympearson) [inline test/node/bson_undefined.test.ts:26]:
```suggestion
        const jsObject = { a: undefined };
```

Is there a reason we can't use an object literal?

File: test/node/bson_undefined.test.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        const bytes = BSON.serialize(jsObject, { ignoreUndefined: false });
+        expect(bytes).to.have.lengthOf(8);
+        expect(bytes).to.have.own.property('4', BSON_DATA_NULL);
```
Comment (baileympearson) [inline test/node/bson_undefined.test.ts:29]: What is happening in this assertion?  Does `haveOwnProperty` check for index access if the element is a Buffer?

Could we use parseToElementsToArray to re-parse the serialized bytes and confirm the serialized bytes contain one BSON null?

File: test/node/bson_undefined.test.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        const jsObject = BSON.deserialize(bsonDocWithUndefined);
+        const bytes = BSON.serialize(jsObject);
+        expect(bytes).to.deep.equal(Uint8Array.of(5, 0, 0, 0, 0));
+      });
+    });
+  });
+
+  describe('when undefined appears inside an array', function () {
+    describe('when ignoreUndefined is set to true', function () {
+      it('does not ignore undefined values', function () {
```
Comment (baileympearson) [inline test/node/bson_undefined.test.ts:52]:
```suggestion
      it('serializes undefined values as null', function () {
```

May as well be consistent with our description

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: test/node/bson_undefined.test.ts (hunk 1 above)
Comment (nbbeeken) [reply test/node/bson_undefined.test.ts]: good point, fixed

File: test/node/bson_undefined.test.ts (hunk 2 above)
Comment (nbbeeken) [reply test/node/bson_undefined.test.ts]: good use of parseToElements

File: test/node/bson_undefined.test.ts (hunk 3 above)
Comment (nbbeeken) [reply test/node/bson_undefined.test.ts]: fixed

**Review by baileympearson — APPROVED** [review]

**Pull request merged**